### ISSUE TITLE
Null returned from OAuth2Client.RequestClientCredentialsAsync()?

### ISSUE THREAD

**Issue by donmiller714** [IdentityServer/IdentityServer3#1335]
Hopefully I'll get all the terminology right.... I think I found an issue but I'm not sure what component because it's all very new to me.

...I'm not even sure this is an issue with IdentityServer3, but it seems it might be so I'm posting just in case...

Following the example as shown in the documentation at http://s3.amazonaws.com/dnr/dotnetrocks_1089_identity.mp3, and the matching sample code, I was unable to get an access token to the Web API.  I tried with the source code from from the dev branch. The problem was that I was getting null back from the OAuth2Client.RequestClientCredentialsAsync() regardless of what I did. Not knowing whether it was on my end or not I exhaustively checked my code and it all looked right.

Finally, I swapped out the versions of IdentityServer3 for the (older?) Thinktecture.IdentityServer3 versions and it worked first time.

It might also have been caused by a mismatch in my MVC client -- that project was using the 'Thinktecture' versions while the server was using the 'IdentityServer' version.

In either case, thought I'd let someone know.

thanks!

**Comment by leastprivilege** [comment]
any more insights?

**Comment by donmiller714** [comment]
Hi Dominick,

Everything seems to be working fine but frankly it's very complex to
configure (for me anyway because of my limited experience in this area) and
I'm sure it will need a revisit before production.However for now it's
fantastic and I've been able to move forward with one of the most important
aspects of my application. Thanks for all the effort you and the team have
put into this project!

There is something I'd like to be able to do with it though, I just don't
know how or even if it's possible. I'm using an MVC client and can get my
user authenticated and authorized for client methods just fine. I can also
call into into my web api with the user's claims as you show in the sample
(I added preferred_username to my user claims so I even get that too; I use
the username to write out to the database who made the last change to the
record). What I'd like to do is call into a second web api from the first,
with my user credentials passed into the second web api, but all I seem to
be able to do is get a machine-to-machine token from the first web api and
use that to cal into the second web api. That pretty much means that all I
can (or should) do is make read only calls to the second web api. It would
be nice to be able to update a record and have the user's username in there
someplace.

To be perfectly honest, it's probably not a good design to allow one web
api to execute write operations on a second web api, and I can forget about
atomic transactions between the two api's in that scenario, so I've
restricted myself to read-only operations on the second web api. However
there may be a future scenario where I want to update a record in the
second web api from the first.

thoughts?

Thanks!
Don
